Migrate ShopParameters->General page

rokaszygmantas opened the pull request:
<!-- Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows with the necessary information: -->

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Migration of ShopParameters->General page to Symfony
| Type?         | new feature
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | n/a
| How to test?  | Compare with the legacy AdminPreferences controller - it should work the same way.

<!-- Click the form's "Preview button" to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/8883)
<!-- Reviewable:end -->

rokaszygmantas committed:
Migrate AdminPreferencesController to Symfony

rokaszygmantas committed:
Enable AdvancedParameters->Multistore tab when Multistore option is enabled

rokaszygmantas committed:
Add missing licence

rokaszygmantas committed:
Migrate controller link generation

rokaszygmantas committed:
Remove old controller and templates

mickaelandrieu commented on src/PrestaShopBundle/Controller/Admin/Configure/ShopParameters/PreferencesController.php:
```diff
@@ -0,0 +1,151 @@
+
+        if (0 === count($saveErrors)) {
+            /** @var EntityManager $em */
+            $em = $this->get('doctrine.orm.entity_manager');
```
can you extract this logic into his own service? the idea is "no business logic in controllers" so later we can unit test every class you create without rely on http layer :)

mickaelandrieu commented on src/PrestaShopBundle/Form/Admin/ShopParameters/General/PreferencesType.php:
```diff
@@ -0,0 +1,197 @@
+            $builder
+                ->add('enable_ssl', ChoiceType::class, array(
+                    'choices_as_values' => true,
+                    'choices'  => array(
```
Note for later: we have a SwitchType that is going to be merged soon => https://github.com/PrestaShop/PrestaShop/pull/8705

So for now, it's good but later I'll update your work and you will have to rely on SwitchType for every "ON/OFF" input :)

mickaelandrieu commented on src/PrestaShopBundle/Form/Admin/ShopParameters/General/PreferencesType.php:
```diff
@@ -0,0 +1,197 @@
+            ->add('shop_activity', ChoiceType::class, array(
+                'choices_as_values' => true,
+                'choices'  => array(
+                    '-- Please choose your main activity --' => 0,
```
what you are doing here is create a group => https://symfony.com/doc/3.4/reference/forms/types/choice.html#grouping-options

mickaelandrieu commented:
Also you forgot to remove the related unit test on the old controller ;)

> 1) Tests\PrestaShopBundle\Controller\ControllerTest::itShouldRunTheTestedController with data set #1 ('AdminPreferencesController')

rokaszygmantas commented:
Thanks for feedback @mickaelandrieu, I'll update my PR :)

mickaelandrieu commented:
@rokaszygmantas you'll also need to rebase your branch :+1:

alegout commented:
Review status: 0 of 16 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[src/PrestaShopBundle/Form/Admin/ShopParameters/General/PreferencesType.php, line 152 at r1](https://beta.reviewable.io/reviews/prestashop/prestashop/8883#-L8Xl3-iIZ3DkFlEWvOt-r1-152:-L8e1_VSEPeK0w4dhw67:b-ydc2ss) ([raw file](https://github.com/prestashop/prestashop/blob/85bb4b201f6db73b49930a477b461b3ae1978130/src/PrestaShopBundle/Form/Admin/ShopParameters/General/PreferencesType.php#L152)):*
<details><summary><i>Previously, mickaelandrieu (Mickaël Andrieu) wrote…</i></summary><blockquote>

what you are doing here is create a group => https://symfony.com/doc/3.4/reference/forms/types/choice.html#grouping-options
</blockquote></details>

Why not using https://symfony.com/doc/current/reference/forms/types/choice.html#placeholder?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/prestashop/prestashop/8883)*
<!-- Sent from Reviewable.io -->

mickaelandrieu commented:
> Why not using https://symfony.com/doc/current/reference/forms/types/choice.html#placeholder?

it make sense only if you want to send a default value: I think the example from docs is disturbing as it will set a value by default and not a placeholder (like a text input placeholder which is not sent to the server but only help the user).

alegout commented:
Oh ok. 
For me the default value is data option, here they talk about "Add an empty value with "Choose an option" as the text"

mickaelandrieu commented:
~~@alegout if you prefer `placeholder` option, I'm fine with it. We only need to ensure this option is not selectable and not sent to the server :+1:~~ 

Erratum: `placeholder` is the right answer, thanks @alegout :)

rokaszygmantas committed:
Remove legacy preferences controller test

rokaszygmantas committed:
Change choice form types with yes/no values to switch type

rokaszygmantas committed:
Move shop activity placeholder text to choice type's placeholder

rokaszygmantas committed:
Move tabs business logic out of controller to a service

rokaszygmantas commented:
@mickaelandrieu I've updated my PR with these changes:
- Removed legacy controller test
- Changed ChoiceType to SwitchType, since the SwitchType PR was already merged
- Added a ChoiceType placeholder for shop activity field, as discussed above
- Moved business logic out of controller to a service

Waiting for your feedback :)

mickaelandrieu commented on src/Adapter/Preferences/PreferencesConfiguration.php:
```diff
@@ -0,0 +1,115 @@
+    public function getConfiguration()
+    {
+        return [
+            'enable_ssl' => (bool) $this->configuration->get('PS_SSL_ENABLED'),
```
```diff
- (bool) $this->configuration->get('PS_SSL_ENABLED')
+ $this->configuration->getBoolean('PS_SSL_ENABLED')
```

See https://github.com/PrestaShop/PrestaShop/pull/8705

rokaszygmantas replied:
@mickaelandrieu I've read the conversation on the PR, if the value is not bool here it will display incorrect style of the switch, or did I misunderstand something?

Quetzacoalt91 replied:
I guess you missed the point in the second line. @mickaelandrieu is suggesting to use `getBoolean` instead of a `(bool)` cast.
The discussion on the other PR was about these forgotten calls. :)

rokaszygmantas replied:
oh right I missed that, thanks @Quetzacoalt91 :)

mickaelandrieu commented on src/Adapter/Preferences/PreferencesConfiguration.php:
```diff
@@ -0,0 +1,115 @@
+     */
+    public function validateConfiguration(array $configuration)
+    {
+        $resolver = new OptionsResolver();
```
use isset() checks instead See https://github.com/PrestaShop/PrestaShop/pull/8368

Sorry we're moving fast this week ^^

mickaelandrieu commented on src/PrestaShopBundle/Controller/Admin/Configure/ShopParameters/PreferencesController.php:
```diff
@@ -0,0 +1,145 @@
+    /**
+     * @param Request $request
+     * @param FormInterface|null $form
+     *
```
Use `@Template` annotation instead, see https://github.com/PrestaShop/PrestaShop/pull/8895

mickaelandrieu commented on src/PrestaShopBundle/Controller/Admin/Configure/ShopParameters/PreferencesController.php:
```diff
@@ -0,0 +1,145 @@
+        // SSL URI is used for the merchant to check if he has SSL enabled
+        $sslUri = 'https://'.$toolsAdapter->getShopDomainSsl().$request->getRequestUri();
+
+        $twigValues = array(
```
using the annotation you can return the array directly

mickaelandrieu commented on src/PrestaShopBundle/Service/Tab/TabStatus.php:
```diff
@@ -0,0 +1,47 @@
+use PrestaShopBundle\Entity\Repository\TabRepository;
+use PrestaShopBundle\Entity\Tab;
+
+class TabStatus
```
class name make me think it's an entity or an object model... how about declare a TabRepository if it doesn't exist and implement changeStatusByClassName inside it? The implementation of the function is OK :+1:

mickaelandrieu commented:
@rokaszygmantas good job, a lot of minor changes according to some changes we have merged and others we are merging.

This is really close to LGTM status: congrats :)

rokaszygmantas committed:
Change template to annotation in preferences controller

rokaszygmantas committed:
Remove poorly named tab service, using tab repository instead

rokaszygmantas committed:
Use getBoolean() instead of bool cast on configuration values

rokaszygmantas committed:
Remove options resolver from configuration validation

rokaszygmantas commented:
I've updated my PR with the small fixes that you requested, @mickaelandrieu :+1:

mickaelandrieu commented:
@rokaszygmantas just noticed that the legacy controller is used by AdminStatsTabControllerCore, can you "unremove" it please? The unit tests should works.

rokaszygmantas committed:
Restore legacy preferences controller

rokaszygmantas commented:
Nice catch @mickaelandrieu, thanks. Restored the legacy controller :)

marionf commented:
Hello @rokaszygmantas 

I have some feedbacks for you:

- The placeholder " -- Please choose your main activity --" should be translated, the translation is available on crowdin 
![capture du 2018-04-05 15-39-35](https://user-images.githubusercontent.com/13449658/38370369-4219bdd4-38ea-11e8-820a-ec43482733b0.png)

- When I change something & click on save I have to choose an activity, it wasn't the case before
![capture du 2018-04-05 16-00-14](https://user-images.githubusercontent.com/13449658/38370557-adfb1dea-38ea-11e8-9178-7a4a5df084e3.png)

mickaelandrieu commented:
Hello @rokaszygmantas,

use the translator for the placeholder (you can extend TranslatorAwareType) and put the required value to false.

And then ping me, I'll validate this very last comment :+1:

rokaszygmantas committed:
Make shop activity field not required

rokaszygmantas committed:
Make shop activity placeholder translatable

rokaszygmantas committed:
Remove unused namespace

rokaszygmantas commented:
@mickaelandrieu ping :) added small fixes

marionf commented:
@rokaszygmantas 
Thank you for the fixes :)
When I change a parameter (no matter which one) and click on save, it says update successful but it's not saved

rokaszygmantas committed:
Fix shop activity to not prevent saving other settings

rokaszygmantas commented:
@marionf thanks for the feedback! Apparently if you did not select anything in shop activity it would not save other settings as well. This should be fixed with my latest commit :)

marionf commented:
Thank you @rokaszygmantas, it's ok for me :)

mickaelandrieu commented:
Thanks @rokaszygmantas  and @marionf  for QA approval :+1:
